enhancements/operator-conditions.md

```diff
@@ -0,0 +1,188 @@
+
+As shown in the example above, by allowing operators to report complex condition and influence OLM's management, the overall user experience could be improved for both operator authors and OLM users.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate conditions that influence OLM behavior.
+- Allow the list of "OLM Supported Conditions" to grow in the future.
+- Update OLM to support the "Upgrade Readiness" condition defined above.
+- Allow operators to communicate additional information to admins without influencing OLM behavior.
+- Preserve existing behavior if the operator does not take advantage of this feature.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want operators managed by OLM to be able to communicate complex states that OLM cannot infer on its own.
+
+#### Story 2
+
+As an operator author, I want operators managed by OLM to be able to communicate whether or not they are ready to be upgraded.
+
+## Design Details
+
+### OperatorHealth CRD
+
+OLM will create a channel for operators to communicate complex conditions by introducing the `OperatorHealth` [CustomResourceDefinition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+When an operator is installed, OLM will create an `OperatorHealth` CustomResource (CR) for the operator. The operator should only be concerned with updating the status of the `OperatorHealth` that it owns and will have its RBAC scoped so it may only GET, LIST, or UPDATE the `OperatorHealth` CR it is associated with.
+
+The `OperatorHealth` CR will be created with the [aggregation label](https://github.com/openshift/enhancements/blob/master/enhancements/olm/simplify-apis.md#component-selection) introduced for the `OperatorStatus` CRD that will associate it with a specific operator. When OLM deploys the operator, an additional environment variable will be appended to the deployment. Operators will then be able to use this label along with a library provided by the [operator-sdk](https://github.com/operator-framework/operator-sdk) team to set [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions) on the CR to communicate complex states to OLM. For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:
```

Comment:
do operators automatically get list/update for this new kind? would that be restricted by name?